src/formats/pnm.c

```diff
@@ -0,0 +1,222 @@
+struct pnm_iter {
+    const uint8_t* data;
+    size_t pos;
+    size_t size;
+    jmp_buf jmp;
+};
+
+// Error conditions as passed to longjmp
+#define ERR_EOF     1
+#define ERR_RANGE   2
+#define ERR_NONDIG  3
+
+// Read a character, ignoring comments
+static uint8_t pnm_readc(struct pnm_iter* it)
+{
+    if (it->pos == it->size) longjmp(it->jmp, ERR_EOF);
+    if (it->data[it->pos] == '#') {
+        do {
+            ++it->pos;
+        } while (it->pos < it->size && it->data[it->pos] != '\n'
+                 && it->data[it->pos] != '\r');
+        if (it->pos == it->size) longjmp(it->jmp, ERR_EOF);
+    }
+    return it->data[it->pos++];
+}
+
+// Read an integer, ignoring leading whitespace
+static size_t pnm_readuint(struct pnm_iter* it)
+{
+    uint8_t c;
+    do {
+        c = pnm_readc(it);
+    } while (c == ' ' || c == '\t' || c == '\n' || c == '\r');
+    if (c < '0' || c > '9') longjmp(it->jmp, ERR_NONDIG);
+    size_t val = 0;
+    do {
+        uint8_t d = c - '0';
+        if (val > SIZE_MAX / 10) longjmp(it->jmp, ERR_RANGE);
+        val *= 10;
+        if (val > SIZE_MAX - d) longjmp(it->jmp, ERR_RANGE);
+        val += d;
+        c = pnm_readc(it);
+    } while (c >= '0' && c <= '9');
+    --it->pos;
+    return val;
+}
+
+// PNM loader implementation
+enum loader_status decode_pnm(struct image* ctx, const uint8_t* data,
+                              size_t size)
+{
+    bool plain;
+    int type;
+    struct pnm_iter it;
+    size_t width, height, maxval;
+
+    if (size < 2 || data[0] != 'P' || data[1] < '1' || data[1] > '6') {
+        return ldr_unsupported;
+    }
+    plain = data[1] < '4';
+    type = (data[1] - '1') % 3;
+    it.data = data;
+    it.pos = 2;
+    it.size = size;
+
+    switch (setjmp(it.jmp)) {
+        case 0:
+            break;
+        case ERR_EOF:
+            image_print_error(ctx, "unexpected end of image");
+            image_free_frames(ctx);
+            return ldr_fmterror;
+        case ERR_RANGE:
+            image_print_error(ctx, "integer too large");
+            image_free_frames(ctx);
+            return ldr_fmterror;
+        case ERR_NONDIG:
+            image_print_error(ctx, "digit expected");
+            image_free_frames(ctx);
+            return ldr_fmterror;
+        default:
+            image_print_error(ctx, "internal error");
+            return ldr_fmterror;
+    }
+
+    width = pnm_readuint(&it);
+    height = pnm_readuint(&it);
+    if (type) {
+        maxval = pnm_readuint(&it);
+        if (!maxval || maxval > 65535) {
+            image_print_error(ctx, "invalid maxval %zu", maxval);
+            return ldr_fmterror;
+        }
+    }
+    uint8_t c = pnm_readc(&it);
+    if (c != ' ' && c != '\t' && c != '\n' && c != '\r') {
+        image_print_error(ctx, "single whitespace character expected");
+        return ldr_fmterror;
+    }
+    if (!image_create_frame(ctx, width, height)) {
+        return ldr_fmterror;
+    }
+
+    if (plain) {
```

Comment:
Will do! There's also a question of how much splitting is best; one could write a function for both plain and raw formats, or 6 functions for plain and raw versions of each of PBM, PGM, and PPM; splitting more involves more code duplication but may also be more efficient since there will be fewer branches in loops.